// @stdlib/stats/incr/wstdev/test/test.js
/**
* @license Apache-2.0
*
* Copyright (c) 2025 The Stdlib Authors.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*    http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

'use strict';

// MODULES //

var tape = require( 'tape' );
var EPS = require( '@stdlib/constants/float64/eps' );
var abs = require( '@stdlib/math/base/special/abs' );
var isnan = require( '@stdlib/math/base/assert/is-nan' );
var sqrt = require( '@stdlib/math/base/special/sqrt' );
var incrwstdev = require( './../lib' );


// TESTS //

tape( 'main export is a function', function test( t ) {
	t.ok( true, __filename );
	t.strictEqual( typeof incrwstdev, 'function', 'main export is a function' );
	t.end();
});

tape( 'the function returns an accumulator function', function test( t ) {
	t.equal( typeof incrwstdev(), 'function', 'returns a function' );
	t.end();
});

tape( 'the function returns an accumulator function (known mean)', function test( t ) {
	t.equal( typeof incrwstdev( 3.0 ), 'function', 'returns a function' );
	t.end();
});

tape( 'the function throws an error if provided a non-numeric value', function test( t ) {
	var values;
	var i;

	values = [
		'5',
		true,
		false,
		null,
		void 0,
		[],
		{},
		function noop() {}
	];
	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[i] ), TypeError, 'throws an error when provided '+values[i] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			incrwstdev( value );
		};
	}
});

tape( 'the weighted corrected sample standard deviation  is `null` until at least 1 datum has been provided (known mean)', function test( t ) {
	var acc;
	var s;

	acc = incrwstdev( 3.0 );

	s = acc();
	t.equal( s, null, 'returns null' );

	s = acc( 3.0 );
	t.notEqual( s, null, 'does not return null' );

	s = acc();
	t.notEqual( s, null, 'does not return null' );

	t.end();
});

tape( 'the accumulator function incrementally computes a weighted standard deviation', function test( t ) {
	var wMeanPrev;
	var expected;
	var wVarSum;
	var actual;
	var delta;
	var dataX;
	var dataW;
	var xwSum;
	var wMean;
	var wSum;
	var acc;
	var tol;
	var N;
	var x;
	var w;
	var i;

	dataX = [ 2.0, 3.0, 2.0, 4.0, 3.0, 4.0 ];
	dataW = [ 1.0, 2.0, 0.1, 1.8, 9.9, 3.6 ];
	N = dataX.length;

	acc = incrwstdev();

	wMeanPrev = 0.0;
	wVarSum = 0.0;
	xwSum = 0.0;
	wMean = 0.0;
	wSum = 0.0;
	for ( i = 0; i < N; i++ ) {
		x = dataX[ i ];
		w = dataW[ i ];
		xwSum += x * w;
		wSum += w;
		wMean = xwSum / wSum;
		wVarSum += w * (x - wMeanPrev) * (x - wMean);
		wMeanPrev = wMean;
		expected = sqrt( wVarSum / wSum );
		actual = acc( x, w );
		delta = abs( actual - expected );
		tol = EPS * abs( expected );
		t.ok( delta <= tol, 'within tolerance. x: ' + x + '. Value: ' + actual + '. Expected: ' + expected + '. Tolerance: ' + tol + '.' );
	}
	t.end();
});

tape( 'if not provided arguments, the accumulator function returns the current weighted standard deviation', function test( t ) {
	var expected;
	var actual;
	var delta;
	var dataX;
	var dataW;
	var tol;
	var acc;
	var i;

	expected = 2.0;
	dataX = [ 2.0, 4.0, 6.0, 8.0 ];
	dataW = [ 1.0, 2.0, 3.0, 4.0 ];
	acc = incrwstdev();
	for ( i = 0; i < dataX.length; i++ ) {
		acc( dataX[ i ], dataW[ i ] );
	}
	actual = acc();
	delta = abs( actual - expected );
	tol = EPS * abs( expected );
	t.ok( delta <= tol, 'returns the current accumulated standard deviation within tolerance' );
	t.end();
});

tape( 'if not provided an input value, the accumulator function returns the current weighted corrected sample standard deviation (known mean)', function test( t ) {
	var dataX;
	var dataW;
	var acc;
	var i;

	dataX = [ 2.0, 4.0, 6.0, 8.0 ];
	dataW = [ 1.0, 2.0, 3.0, 4.0 ];
	acc = incrwstdev( 6.0 );
	for ( i = 0; i < dataX.length; i++ ) {
		acc( dataX[ i ], dataW[ i ] );
	}
	t.equal( acc(), 2.0, 'returns the current accumulated weighted corrected sample standard deviation' );
	t.end();
});

tape( 'if not provided a weight, the accumulator function returns `NaN`', function test( t ) {
	var acc = incrwstdev();
	t.equal( isnan( acc( 2.0 ) ), true, 'returns NaN' );
	t.equal( isnan( acc( 3.14 ) ), true, 'returns NaN' );
	t.end();
});

tape( 'if provided `NaN` for either a value or a weight, the accumulator function returns `NaN`', function test( t ) {
	var acc = incrwstdev();
	t.equal( isnan( acc( 2.0, NaN ) ), true, 'returns NaN' );
	t.equal( isnan( acc( 3.14, NaN ) ), true, 'returns NaN' );

	acc = incrwstdev();
	t.equal( isnan( acc( NaN, 1.0 ) ), true, 'returns NaN' );
	t.equal( isnan( acc( NaN, 1.0 ) ), true, 'returns NaN' );

	acc = incrwstdev();
	t.equal( isnan( acc( NaN, NaN ) ), true, 'returns NaN' );
	t.equal( isnan( acc( NaN, NaN ) ), true, 'returns NaN' );
	t.end();
});
